### PR TITLE
Add load_file performance benchmark tooling

### PR DESCRIPTION
**Pull request by tatiana**
As part of this PR, I implemented a benchmark tool to support us in analyzing the performance of Astro operators and tasks.
At the moment it is being used to evaluate `load_file`, but in the next PR, it will be generalized to be used with other operators.

Relates to issue #108 

Considering the chunk sizes are set to 2,000, the performance of the operators was:

### Database: BigQuery

| database   | dataset   | total_time   | memory_rss   | memory_shared   | memory_pss   | cpu_time_user   | cpu_time_system   |
|:-----------|:----------|:-------------|:-------------|:----------------|:-------------|:----------------|:------------------|
| bigquery   | few_kb    | 11.28s       | 120.73MB     | 40.91MB         | 118.02MB     | 3.5s            | 1.63s             |
| bigquery   | many_kb   | 18.57s       | 122.39MB     | 40.68MB         | 119.74MB     | 3.55s           | 1.58s             |
| bigquery   | few_mb    | 42.99min     | 140.99MB     | 40.94MB         | 138.21MB     | 20.07s          | 2.27s             |

### Database: postgres

| database   | dataset   | total_time   | memory_rss   | memory_shared   | memory_pss   | cpu_time_user   | cpu_time_system   |
|:-----------|:----------|:-------------|:-------------|:----------------|:-------------|:----------------|:------------------|
| postgres   | few_kb    | 1.73s        | 125.62MB     | 41.89MB         | 122.95MB     | 1.95s           | 1.4s              |
| postgres   | many_kb   | 2.24s        | 138.5MB      | 41.78MB         | 135.95MB     | 2.4s            | 1.38s             |
| postgres   | few_mb    | 1.23min      | 156.85MB     | 41.96MB         | 154.15MB     | 1.09min         | 1.74s             |
| postgres   | many_mb   | 23.94min     | 1.23GB       | 41.95MB         | 1.23GB       | 22.15min        | 31.5s            |

### Database: snowflake

| database   | dataset   | total_time   | memory_rss   | memory_shared   | memory_pss   | cpu_time_user   | cpu_time_system   |
|:-----------|:----------|:-------------|:-------------|:----------------|:-------------|:----------------|:------------------|
| snowflake  | few_kb    | 14.28s       | 136.06MB     | 50.39MB         | 132.96MB     | 3.12s           | 1.48s             |
| snowflake  | many_kb   | 12.42s       | 138.83MB     | 50.22MB         | 135.85MB     | 3.17s           | 1.48s             |
| snowflake  | few_mb    | 7.21min      | 264.67MB     | 50.2MB          | 261.61MB     | 25.95s          | 3.27s             |
| snowflake  | many_mb   | 4.86min      | 975.47MB     | 50.48MB         | 972.4MB      | 41.67s          | 5.03s             |

### CODE REVIEW

**Review by dimberman — CHANGES_REQUESTED** [review]

File: tests/benchmark/dags/evaluate_load_file.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            output_table=table_metadata,
+            chunksize=2000,
+        )
+
+        # aggregate_check(
+        #    table_xcom,
+        #    check="SELECT COUNT(*) FROM {table}",
+        #    equal_to=dataset_rows
+        # )
+        # table_xcom.set_downstream(aggregate_check)
```
[the remark this answers is not in the document]
Comment (tatiana) [reply tests/benchmark/dags/evaluate_load_file.py:58]: Yes, I believe once the operator is working again, we should re-enable these lines. I believe it is relevant to make sure the load worked as expected for these larger datasets.

File: tests/benchmark/dags/evaluate_load_file.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    dag_name = f"load_file_{dataset_name}_into_{database_name}"
+    table_name = Path(dataset_path).stem
+
+    with DAG(dag_name, schedule_interval=None, start_date=START_DATE) as dag:
+        table_metadata = Table(table_name=table_name, **table_args)
+        table_xcom = aql.load_file(
+            path=dataset_path,
+            task_id="load_csv",
+            output_table=table_metadata,
+            chunksize=2000,
```
[the remark this answers is not in the document]
Comment (tatiana) [reply tests/benchmark/dags/evaluate_load_file.py:50]: I'll play a bit with this and let you know how it goes!

File: tests/benchmark/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        "database": "bigquery",
+        "conn_id": "bigquery"
+      }
+    }
+  ],
+  "datasets": [
+    {
+        "name": "few_kb",
+        "size": "45K",
+        "path":"/tmp/covid_overview.csv",
```
Comment (tatiana) [inline tests/benchmark/README.md:38]: Improve this - allow reproducibility by either adding a command to download these datasets or add them to a public GCS folder.

File: tests/benchmark/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+This is an example of the content of `/tmp/results-2022-02-10T14:36:09.ndjson`:
+
+```
+{"duration": "1.66s", "rss": "125.37MB", "vms": "990.95MB", "shared": "41.73MB", "dag_id": "load_file_few_kb_into_postgres", "execution_date": "2022-02-12 00:33:56.463244+00:00", "revision": "e1fb164"}
+{"duration": "2.28s", "rss": "138.56MB", "vms": "1.0GB", "shared": "41.55MB", "dag_id": "load_file_many_kb_into_postgres", "execution_date": "2022-02-12 00:33:59.808866+00:00", "revision": "e1fb164"}
+{"duration": "1.2min", "rss": "157.17MB", "vms": "1.02GB", "shared": "41.9MB", "dag_id": "load_file_few_mb_into_postgres", "execution_date": "2022-02-12 00:34:03.794995+00:00", "revision": "e1fb164"}
+
+```
+
+## Analyse the results
```
Comment (tatiana) [inline tests/benchmark/README.md:87]: Add information related to the system where we obtained these results.

File: tests/benchmark/run.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    return {key: after_dict[key] - before_dict.get(key, 0) for key in after_dict}
+
+
+def profile(func, *args, **kwargs):
+    def wrapper(*args, **kwargs):
+        process = psutil.Process(os.getpid())
+        # metrics before
+        memory_full_info_before = process.memory_full_info()._asdict()
+        cpu_time_before = process.cpu_times()._asdict()
+        io_counters_before = process.io_counters()._asdict()
```
Comment (tatiana) [inline tests/benchmark/run.py:35]: MacOS doesn't have this :cry:

File: tests/benchmark/config-sample.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        "name": "many_kb",
+        "size": "472K",
+        "path": "/home/tati/Data/artist_data.csv",
+        "rows": 3532
+    },
+    {
+        "name": "few_mb",
+        "size": "20M",
+        "path": "/home/tati/Data/title_ratings.csv",
```
Comment (tatiana) [inline tests/benchmark/config-sample.json:43]: At the moment this doesn't support relative paths - do we want to do this?

File: tests/benchmark/analyse.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    # format data
+    mean_by_dag["database"] = mean_by_dag.dag_id.apply(
+        lambda text: text.split("into_")[-1]
+    )
+    mean_by_dag["dataset"] = mean_by_dag.dag_id.apply(
+        lambda text: text.split("load_file_")[-1].split("_into")[0]
+    )
+
+    mean_by_dag["memory_pss"] = mean_by_dag.memory_full_info_pss.apply(
```
Comment (tatiana) [inline tests/benchmark/analyse.py:52]: In MacOS psutil does not expose PSS - so we should make this optional.

File: tests/benchmark/analyse.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        lambda value: format_bytes(value)
+    )
+    mean_by_dag["memory_rss"] = mean_by_dag.memory_full_info_rss.apply(
+        lambda value: format_bytes(value)
+    )
+    mean_by_dag["memory_shared"] = mean_by_dag.memory_full_info_shared.apply(
```
Comment (tatiana) [inline tests/benchmark/analyse.py:58]: In MacOS `psutil` also does not expose `memory_shared`

File: tests/benchmark/config-sample.json (hunk 6 above)
Comment (tatiana) [reply tests/benchmark/config-sample.json:43]: It can be an improvement for the future!

**Review by dimberman — APPROVED** [review]

**Pull request merged**